### ISSUE TITLE
Kindle Oasis physical page turn buttons do not work when using KOReader

### ISSUE THREAD

**Issue by homeceo43** [koreader/koreader#2173]
* KOReader version: koreader-kindle-arm-linux-gnueabi-v2015.11-533-g1d2150c.zip [edit: I have this version installed instead of the stable version because the stable version crashes on the Kindle Oasis]
* Device: Kindle Oasis, firmware 5.8.1

1st issue:
The physical page turn buttons do not work while using KOReader.

2nd issue:
The Kindle Oasis has the accelerometer feature where you can flip the Oasis upside down to hold in your left hand, the screen flips orientation and the physical page buttons can be used with left hand.

The screen does flip orientation, the book flips orientation, KOReader menus also flip orientation (so they are not upside down when holding the reader in left hand) however KOReader still thinks it is right side up tap zone-wise -- the tap zones do not flip orientation. Tap the left side of the screen pages forward; tap right side of screen (or center of screen) pages back. Tap top of screen and page number disappears, tap top of screen again page number reappears (page number is what I have showing only, not the other stuff or progress bar). Tap the bottom of the screen brings up the KOReader menu, but tapping anything in the menu just makes the menu disappear.

[edit: I also have a Kindle Voyage, firmware 5.8.1, stable version of KOReader installed; the PagePress buttons to page forward and page backward do function correctly.]

**Comment by Hzj-jie** [comment]
Similar issue on Kindle Voyage with firmware 5.8.1. Page down button works fine (good luck), but page up won't.

**Comment by houqp** [comment]
@Hzj-jie you have a KV now? Can you paste the log for those keycodes here?

**Comment by chrox** [comment]
@Hzj-jie You need to launch Koreader with KPVBooklet in order to use the Whisper Touch function . When launched with KUAL Koreader can only get the keycode of page down but never page up as KUAL Kindlet does not request Whisper Touch support in the appreg.db.

Check out [this](https://github.com/koreader/kpvbooklet/blob/master/install.ffs#L48-L52) for more info:
```
	WHISPERTOUCH=`sqlite3 /var/local/appreg.db .dump|grep "'whisper-touch','supported'"|wc -l`
	if [ "$WHISPERTOUCH" \> 0 ]; then
		progress 50 "Enable whisper touch for Kindle Voyage"
		sqlite3 /var/local/appreg.db < whispertouch.install.sql
	fi
```

@NiLuJe Maybe KUALBooklet could also add support of this.

**Comment by NiLuJe** [comment]
@chrox: Yeah, I was considering that... Speaking of, do you have any idea what the asr & grip-suppression entries might be for?

**Comment by homeceo43** [comment]
Just to add: I use KPVBooklet 0.6.4 to launch KOReader on my Kindle Oasis 5.8.1.

**Comment by Hzj-jie** [comment]
@HouQP, yep, I got it for a while, but I have just jailbroken it.
@chrox, emmm, that's interesting. I have installed KPVBooklet, but cannot find it anywhere. So I thought the one in KUAL is the right launcher. No matter, if this is not a real trouble.

**Comment by woshilaiba** [comment]
@chrox  in my KO the KPVBooklet is already installed as you pointed out, but still the physical page turn buttons do not work, so strange~~

**Comment by bingo619** [comment]
Hi everyone,
I have just made a pull request for fix of the buttons.
I used the same key code for Voyage, for the input event, as NiLuJe pointed out [](https://github.com/koreader/koreader/issues/2181) we need to get the input by name as 3G model has got an extra input
so I used "/dev/input/by-path/platform-gpiokey.0-event". 

Tested it on my 3g model, all working good.

For the orientation bug, any one got a clue? Really want to fix this as well.

Thanks.

**Comment by NiLuJe** [comment]
@bingo619: Thanks :).

Quick question: Does the physical buttons work when started from KUAL? Or only from KPVBooklet?

As for the orientation thing, it's probably fixable, but someone actually familiar with the orientation code would better be able to point you in the right direction ;).

**Comment by homeceo43** [comment]
Thank you @bingo619 and @NiLuJe for working to find some solution for this.

Kindle Oasis, fw 5.8.1 , wi-fi only
KOReader v2015.11-533-g1d2150c
KPVBooklet 0.6.4

I copied that bit of @bingo619's code to the appropriate file in KOReader on my device (frontend/device/kindle/device.lua), but whether I use KUAL or KPVBooklet (my preference) to launch KOReader, the physical page turn buttons still do not work for me. My Oasis is wi-fi only, not 3g, if that makes a difference.

I don't know if I had edited the device.lua file properly though; on your code the line starts at 342. On the file on my device, the line is 337. (Use notepad++ to view and edit).

**Comment by bingo619** [comment]
Sorry guys, I missed one line of code [code](https://github.com/bingo619/koreader/blob/master/frontend/device/kindle/device.lua#L112), I should have tested that before pushing, so sorry.
@NiLuJe yes, it does work on both.

**Comment by homeceo43** [comment]
Yes! I added that last bit of code and now the page turn buttons work in KOReader when launched from either KPVBooklet or KUAL.

Awesome, and thank you @bingo619!

**Comment by houqp** [comment]
Thanks @bingo619 for the fix! For the orientation support, is there anyway to capture the orientation change event from the kernel or kindle framework? 

If so, we just need to create a custom event handler for the orientation change event like this one: https://github.com/koreader/koreader/blob/master/frontend/ui/uimanager.lua#L50

Then inside the handler flip the screen and switch the key mapping.

**Comment by bingo619** [comment]
@houqp Thanks for pointing out the direction, I will take a look at if the we can get the event.

**Comment by bingo619** [comment]
@homeceo43 Just pushed a commit to fix the orientation bug.

An known bug is that the KO dos't know the initial orientation of the device, so if you opened KO upside down, it still doesn't work.

@everyone Does anybody know how to get a initial orientation? or potentially we can make it Portrait even if we opened it Upside Down by sending a fake event?

**Comment by houqp** [comment]
Is the information not exported in /sys or /proc?

**Comment by bingo619** [comment]
Can't find anything relating to this in /sys nor /proc.

**Comment by NiLuJe** [comment]
A lipc property, perhaps? ```lipc-probe -av``` is your friend ;).

**Comment by vincent-inspectionapps** [comment]
@NiLuJe Thanks. Will try it later. Sorry about using two github accounts, it is a bit confusing, isn't it? :)

**Comment by houqp** [comment]
LOL, I was about to ask whether you own both of the accounts...

**Comment by bingo619** [comment]
@NiLuJe  @houqp Thanks NiLuJe, I can get the orientation from the lipc property now.
`
    local lipc = require("liblipclua")
    if not lipc then
        DEBUG("could not load liblibclua")
        return false
    end
    local lipc_handle = lipc.init("com.github.koreader.screen")
    if not lipc_handle then
        DEBUG("could not get lipc handle")
        return false
    end

    local orientation_code = lipc_handle:get_string_property("com.lab126.winmgr", "accelerometer")
`
But I am not sure how to initial the orientation to the device.

I have added a new property to the device

`    -- startup
    isAlwaysPortrait = no,
    -- needs full screen refresh when resumed from screensaver?
    needsScreenRefreshAfterResume = yes,

    native_rotation_mode = 0
}
`
So I can get it from framebuffer.

There are 3 orientation ralative codes which are blitbuffer_rotation_mode, native_rotation_mode, cur_rotation_mode, what exactly are there for? which one should i set?

**Comment by bingo619** [comment]
I guess I see the problem now, it sets blitbuffer_rotation_mode, native_rotation_mode, cur_rotation_mode all to 0 no matter which orientation you hold when you start KO.

I guess the easier way for now is that set the blitbuffer_rotation_mode = 2 if you hold it upside down?  You will see the reader view upside down, but if you flip it, all will be all working fine.

**Comment by bingo619** [comment]
Forget what i just said, I guess we need to do something in the base.

Anyone can help this please.

**Comment by houqp** [comment]
Sorry for the late response, finally got some time to catch up on the github issues >_<.

What orientation_code did you get from lipc? Are they consistent with the notation documented in framebuffer.lua?

So IIRC, in framebuffer.lua, you only need to initialize self.native_rotation_mode and assign the same value to self.cur_rotation_mode. You should be able to leave blitbuffer_rotation_mode untouched.

**Comment by bingo619** [comment]
Thanks for this.

See #2209

**Comment by chrox** [comment]
Implemented in #2209, close this now.
